Cython/Compiler/ExprNodes.py

```diff
@@ -3981,14 +3988,15 @@ def analyse_as_type(self, env):
 
     def analyse_pytyping_modifiers(self, env):
         # Check for declaration modifiers, e.g. "typing.Optional[...]" or "dataclasses.InitVar[...]"
+        # `typing.Optional` is used for all variants of modifiers representing Optional type (Optional[T], Union[T, None])
         # TODO: somehow bring this together with TemplatedTypeNode.analyse_pytyping_modifiers()
         modifiers = []
         modifier_node = self
         while modifier_node.is_subscript:
             modifier_type = modifier_node.base.analyse_as_type(env)
             if (modifier_type and modifier_type.python_type_constructor_name
                     and modifier_type.modifier_name):
-                modifiers.append(modifier_type.modifier_name)
+                modifiers.append('typing.Optional' if modifier_type.allows_none() else modifier_type.modifier_name)
```

Comment:
What happens if a user writes `Optional[Union[int, None]]`? We'd probably get a duplicate `Optional` in that case, right? Although I'd guess that that won't even hurt us…